File: Algorithm/Sorting Algorithm/0075 Sort Colors.py
class Solution:
  def sortColors(self, nums: List[int]) -> None:
    """
    Do not return anything, modify nums in-place instead.
    """
    # initialize several pointers
    l = 0
    r = len(nums) - 1
    cur = 0
    
    # use two pointers on the two ends
    while (cur <= r):
      # if number is 0, swap with l (to stay on the left)
      if nums[cur] == 0:
        nums[l], nums[cur] = nums[cur], nums[l]
        l += 1
        cur += 1
      # if number is 2, swap with r (to stay on the right)
      elif nums[cur] == 2:
        nums[r], nums[cur] = nums[cur], nums[r]
        r -= 1
      else: 
        cur += 1

File: Algorithm/Sorting Algorithm/0075 Sort Colors.java
class Solution {
  public void sortColors(int[] nums) {
    if (nums.length < 2) return; 
    
    // index pointing to start and end
    int start = 0; 
    int end = nums.length - 1; 
    int i = 0; 
    
    while (i <= end && start < end){
      if (nums[i] == 0){
        nums[i] = nums[start]; 
        nums[start] = 0; 
        start ++; 
        i ++; 
      }
      else if (nums[i] == 2){
        nums[i] = nums[end]; 
        nums[end] = 2; 
        end --; 
      }
      else {
        i ++; 
      }
    }
  }
}